lib/src/main/scala/code/chymyst/jc/ReactionSite.scala

```diff
@@ -287,50 +294,60 @@ private[jc] final class ReactionSite(reactions: Seq[Reaction], reactionPool: Poo
       moleculesPresent.removeFromBag(bm, blockingMolValue)
       if (logLevel > 0) println(s"Debug: $this removed $bm($blockingMolValue) on thread pool $sitePool, now have molecules [${moleculeBagToString(moleculesPresent)}]")
     }
-    blockingMolValue.synchronized {
-      blockingMolValue.replyValue.replyTimedOut = hadTimeout
-    }
+    if (hadTimeout) blockingMolValue.replyValue.replyStatus = ReplyTimedOut
+
   }
 
-  private def emitAndAwaitReplyInternal[T,R](timeoutOpt: Option[Long], bm: BlockingMolecule[T,R], v: T, replyValueWrapper: AbsReplyValue[T,R]): Boolean = {
+  /** Common code for [[emitAndAwaitReply]] and [[emitAndAwaitReplyWithTimeout]].
+    *
+    * @param timeoutOpt Timeout value in nanoseconds, or None if no timeout is requested.
+    * @param bm A blocking molecule to be emitted.
+    * @param v Value that the newly emitted molecule should carry.
+    * @param replyValueWrapper The reply value wrapper for the blocking molecule.
+    * @tparam T Type of the value carried by the blocking molecule.
+    * @tparam R Type of the reply value.
+    * @return Reply status for the reply action.
+    */
+  private def emitAndAwaitReplyInternal[T,R](timeoutOpt: Option[Long], bm: BlockingMolecule[T,R], v: T, replyValueWrapper: AbsReplyValue[T,R]): ReplyStatus = {
+    replyValueWrapper.resetSemaphoreForReply()
+    replyValueWrapper.releaseSemaphoreForReply()
     val blockingMolValue = BlockingMolValue(v, replyValueWrapper)
     emit(bm, blockingMolValue)
     val success =
       BlockingIdle {
         replyValueWrapper.acquireSemaphore(timeoutNanos = timeoutOpt)
       }
-    replyValueWrapper.deleteSemaphore()
     // We might have timed out, in which case we need to forcibly remove the blocking molecule from the soup.
     removeBlockingMolecule(bm, blockingMolValue, !success)
 
-    success
+    val result = replyValueWrapper.replyStatus
+    replyValueWrapper.releaseSemaphoreForReply()
+
+    result
   }
 
   // Adding a blocking molecule may trigger at most one reaction and must return a value of type R.
   // We must make this a blocking call, so we acquire a semaphore (with or without timeout).
   private[jc] def emitAndAwaitReply[T,R](bm: BlockingMolecule[T,R], v: T, replyValueWrapper: AbsReplyValue[T,R]): R = {
-    emitAndAwaitReplyInternal(timeoutOpt = None, bm, v, replyValueWrapper)
+    val result = emitAndAwaitReplyInternal(timeoutOpt = None, bm, v, replyValueWrapper)
```

Comment:
Yes, I would normally do that, but I am using `result` to make an error message below.